### PR TITLE
RFC: Introduce arch_send_cpu_stop to halt secondary cores for fatal errors in SMP systems

### PR DESCRIPTION
**Pull request by quic-lingutla**
On SMP systems, when fatal error occured, there is no common mechanism available to halt secondary cores.
Introduce arch_send_cpu_stop API to halt secondary cores, which gets called in fatal error path (for now arch_system_halt() calls it- can be changed).

As an example, implemented the API for RISC-V architecture, but each ARCH should implement its own mechanism to inform/halt secondary cores.

### CODE REVIEW

**Review by andyross — COMMENTED** [review]
Some notes.  Not opposed to a general API for this, but I'm not sure this is it either.

The reason this has been left undefined (i.e. "owned by the SOC layer") at the arch and kernel levels is that different architectures have very different ideas about how it works.  Some systems can't even stop CPUs individually, some like intel_adsp have to shut them down synchronously (maybe, this is IMHO sort of a wart, but that's how it works right now) in coordination with an external host CPU, etc...

One thing this does skip though, that's really important, is a state predicate.  You can tell a CPU to shut down with this API, but what you can't do is know if it's "definitely off" such that it's safe to turn back on without race conditions.  I suspect that's because you're looking at this as a fatal error handler, but the problem is bigger than that.

File: arch/riscv/core/smp.c [hunk 1]
```diff
@@ -108,25 +130,35 @@ static void ipi_handler(const void *unused)
...
-	MSIP(csr_read(mhartid)) = 0;
+	uint8_t cpu = csr_read(mhartid);
 
-	atomic_val_t pending_ipi = atomic_clear(&cpu_pending_ipi[_current_cpu->id]);
+	while (true) {
+		atomic_val_t pending_ipi = atomic_clear(&cpu_pending_ipi[cpu]);
+
+		MSIP(cpu) = 0U;
+
+		if (!pending_ipi)
```
Comment (quic-egmc) [inline arch/riscv/core/smp.c:140]: add braces

File: arch/riscv/core/smp.c [hunk 2]
```diff
@@ -108,25 +130,35 @@ static void ipi_handler(const void *unused)
...
+			break;
+
+		if (pending_ipi & ATOMIC_MASK(IPI_SCHED))
```
Comment (quic-egmc) [inline arch/riscv/core/smp.c:143]: add braces

File: arch/riscv/core/smp.c [hunk 3]
```diff
@@ -108,25 +130,35 @@ static void ipi_handler(const void *unused)
...
+			csr_clear(mstatus, MSTATUS_IEN);
+			/* perform the flush */
+			z_riscv_flush_local_fpu();
+			/*
+			 * No need to re-enable IRQs here as long as
+			 * this remains the last case.
+			 */
+		}
 #endif
+		if (pending_ipi & ATOMIC_MASK(IPI_CPU_STOP))
```
Comment (quic-egmc) [inline arch/riscv/core/smp.c:158]: add braces

File: arch/riscv/core/smp.c [hunk 4]
```diff
@@ -76,26 +76,48 @@ void z_riscv_secondary_cpu_init(int hartid)
...
+void arch_send_cpu_stop(void)
+{
+	riscv_send_ipi(IPI_CPU_STOP);
+}
+
+static void arch_cpu_stop(void)
+{
+	printk("Stopping CPU: %d\n", _current_cpu->id);
+	while (1)
+		;
```
[the remark this answers is not in the document]
Comment (quic-lingutla) [reply arch/riscv/core/smp.c:118]: Sure. 
But this is called with interrupts disabled right ? (from interrupt handler)

[the remark this answers is not in the document]
Comment (fkokosinski) [reply arch/riscv/core/smp.c:124]: Let's use `arch_cpu_idle` here in the while loop for power saving

**Comment by andyross** [comment]
Still doesn't have state feedback to know a CPU is off (i.e. that it's safe to call arch_start_cpu() again).  As mentioned, without that, this isn't really useful for anything **but** fatal system shutdown.  And given that, I'm not sure it has much value as a general kernel API?

Maybe we should continue to provide that as a private API out of the platform layers and call it from an app-provided sys_fatal_error_handler instead?   Again, CPU lifecycle behavior is really varied between platforms, I don't think many of us have given much thought to a general solution here, and this is really just a shim for the needs of one app on one device?

**Comment by dcpleung** [comment]
TBH, the name `arch_send_cpu_stop` is kind of misleading here. The actual implementation is simply to send IPI. So maybe rename it to `arch_ipi_cpu_stop` or something... or better, just have a generic `arch_ipi()` with some generic value and architecture specific add-ons.

**Pull request closed without merging**